### ISSUE TITLE
Implement whipsaw detection and mitigation strategies

### ISSUE THREAD

**Issue by deanturpin** [deanturpin/lft#9]
## Problem

Getting **whipsawed** - stop losses triggering on normal market noise, causing us to exit positions that would have been winners if given more room to breathe. This is "death by a thousand cuts" - repeatedly taking small -1% losses on trades that reverse shortly after we're stopped out.

## Current Behaviour

- Fixed 1% stop loss for all trades (see [exit_logic_tests.h:22](../include/exit_logic_tests.h#L22))
- Adaptive stop loss implemented but may not be aggressive enough
- No analysis of whether losses are from whipsaw vs genuine reversals
- Spread (2bp stocks, 10bp crypto) eats into the already-thin stop loss buffer

## Symptoms to Watch For in CSV Data

When the stats file is generated, look for:

1. **Low win rate** (<40%) despite profitable calibration
2. **Many small losses** clustering around -1.0% 
3. **Few profitable trades** reaching take profit levels
4. **High loss frequency** - lots of `losses` compared to `wins`

Example red flag:
```csv
2026-01-08 21:11:45,live_trading,mean_reversion,20,15,15,4,11,26.67,45.20,165.00,-119.80,-7.99
```
- 11 losses vs 4 wins → 73% loss rate
- Negative net P&L despite profitable calibration
- Average loss per trade: -$7.99

### 1. Whipsaw Detection Metric

Add to CSV output to track whipsaw rate:

```cpp
// In write_stats_to_file(), add columns:
file << "timestamp,run_type,strategy,signals,trades_executed,trades_closed,"
     << "wins,losses,win_rate,total_profit,total_loss,net_profit,avg_profit,"
     << "whipsaw_count,whipsaw_rate\n";  // NEW

// Calculate whipsaw_count as losses at exactly stop_loss_pct
auto whipsaw_count = count_losses_at_exact_sl(stat);
auto whipsaw_rate = stat.trades_closed > 0 
                     ? (whipsaw_count * 100.0) / stat.trades_closed 
                     : 0.0;
```

**What constitutes a whipsaw:**
- Loss is within 0.1% of exact SL threshold (e.g., -0.95% to -1.05% for 1% SL)
- Trade duration < 5 minutes (too quick to be real reversal)
- Price recovered above entry within 10 minutes of stop (would have been winner)

### 2. Widen Stop Loss Based on Volatility

Already partially implemented via `adaptive_stop_loss()` but may need tuning:

```cpp
// In exit_logic_tests.h, adjust the adaptive multiplier
constexpr double adaptive_stop_loss(double base_sl, double recent_noise) {
    if (recent_noise <= 0.0) return base_sl;
    
    // Currently uses: base_sl * (1.0 + noise)
    // Try more aggressive widening:
    auto volatility_buffer = std::max(1.5, recent_noise * 2.0);  // At least 1.5x base
    return base_sl * volatility_buffer;
}
```

**Example effect:**
- Base SL: 1%
- Recent noise: 0.8% (typical intraday)
- Old adaptive: 1% * 1.8 = 1.8% SL
- New aggressive: 1% * max(1.5, 1.6) = 1.6% SL minimum

### 3. Time-Based Stop Loss Widening

Don't use tight stops immediately after entry - give the trade time to develop:

```cpp
// In exit evaluation logic
auto time_in_trade = current_time - position.entry_time;
auto sl_multiplier = 1.0;

if (time_in_trade < 2min) {
    sl_multiplier = 2.0;  // Double the SL for first 2 minutes
} else if (time_in_trade < 5min) {
    sl_multiplier = 1.5;  // 1.5x SL for 2-5 minutes
}

auto effective_sl = position.stop_loss_pct * sl_multiplier;
```

**Rationale:** Initial entry slippage and spread mean you're starting slightly underwater. Need time for thesis to play out.

### 4. ATR-Based Stop Loss

Use Average True Range instead of fixed percentage:

```cpp
// Add to PriceHistory
double average_true_range(size_t periods = 14) const {
    if (highs.size() < periods + 1) return 0.0;
    
    auto total_tr = 0.0;
    for (auto i = highs.size() - periods; i < highs.size(); ++i) {
        auto tr = std::max({
            highs[i] - lows[i],
            std::abs(highs[i] - prices[i-1]),
            std::abs(lows[i] - prices[i-1])
        });
        total_tr += tr;
    }
    return total_tr / periods;
}

// Use in stop loss calculation
auto atr = history.average_true_range();
auto atr_stop = entry_price - (atr * 2.0);  // 2x ATR stop
auto percent_stop = entry_price * (1.0 - stop_loss_pct);
auto effective_stop = std::min(atr_stop, percent_stop);  // Use wider of the two
```

### 5. Price Confirmation Before Entry

Don't enter immediately when signal fires - wait for price to move in your direction:

```cpp
// For mean reversion buy signals
if (signal.should_buy) {
    // Store the signal, don't execute yet
    pending_signals[symbol] = {signal, current_price, timestamp};
}

// On next cycle, check if price moved favorably
if (pending_signals.contains(symbol)) {
    auto &pending = pending_signals[symbol];
    if (current_price < pending.price * 0.999) {  // 0.1% confirmation move
        // Price moved in our direction, execute entry
        execute_entry(symbol, pending.signal);
    } else if (time_since(pending.timestamp) > 5min) {
        // Too long waiting, cancel signal
        pending_signals.erase(symbol);
    }
}
```

**Trade-off:** Misses some fast moves, but avoids entering right before whipsaw.

### 6. Account for Spread in Entry Logic

Factor spread into profit targets and stop losses:

```cpp
// Current: 2bp spread for stocks means immediate -0.02% drawdown
constexpr auto effective_stop_loss = stop_loss_pct + (stock_spread / 2.0);  // 1.01%
constexpr auto effective_take_profit = take_profit_pct - (stock_spread / 2.0);  // 0.99%

// This makes targets asymmetric but realistic
```

### 7. Skip Low-Volatility Regimes

Already have noise regime detection ([lft.cxx:119-123](../src/lft.cxx#L119-L123)):

```cpp
auto recent_noise = history.recent_noise(20);
auto low_noise_regime = recent_noise > 0.0 and recent_noise < 0.005;  // <0.5%
```

**Use it to block trades:**
```cpp
if (low_noise_regime and signal.should_buy) {
    std::println("{}⏸  BLOCKED: Low volatility regime - likely to whipsaw{}",
                 colour_yellow, colour_reset);
    continue;  // Don't enter when noise is too low
}
```

**Why:** In low volatility, the stop loss is proportionally tighter relative to natural price movement.

## Implementation Priority

**Phase 1 - Diagnostics:**
1. Add whipsaw detection to CSV output
2. Run for a few sessions, collect data
3. Calculate actual whipsaw rate from stats

**Phase 2 - Quick Wins:**
1. Widen adaptive stop loss multiplier
2. Skip low-volatility regime entries
3. Account for spread in effective SL/TP

**Phase 3 - Sophisticated Improvements:**
1. Implement ATR-based stops
2. Add time-based SL widening
3. Price confirmation before entry

## Metrics to Track

After implementing fixes, monitor:
- **Whipsaw rate**: Should drop from >50% to <20%
- **Win rate**: Should improve if stops were too tight
- **Average loss**: May increase slightly (wider stops) but should be offset by more winners
- **Net P&L**: Should improve overall despite some larger losses

## Related

- Current exit logic: [exit_logic_tests.h](../include/exit_logic_tests.h)
- Adaptive stop loss: [exit_logic_tests.h:28-33](../include/exit_logic_tests.h#L28-L33)
- Noise regime detection: [lft.cxx:119-123](../src/lft.cxx#L119-L123)
- Stats logging: [lft.cxx:572-609](../src/lft.cxx#L572-L609)

## Further Reading

Classic trading wisdom:
- "Let your winners run, cut your losses short" - but not TOO short
- "The market can remain irrational longer than you can remain solvent" - need buffer for noise
- "Don't confuse movement with progress" - price wiggle ≠ trend change
